algorithms/linfa-logistic/src/lib.rs

```diff
@@ -421,6 +476,37 @@ fn logistic_loss<F: Float, A: Data<Elem = F>>(
     -yz.sum() + F::cast(0.5) * alpha * params.dot(&params)
 }
 
+/// Compute the log of probabilities, which is `log(softmax(H))`, where H is `X . W + b`. Also
+/// returns `W` without the intercept.
+/// `Y` is the output (n_samples * n_classes), `X` is the input (n_samples * n_features), `W` is the
+/// params (n_features * n_classes), `b` is the intercept vector (n_classes).
+fn multi_logistic_prob_params<F: Float, A: Data<Elem = F>>(
+    x: &ArrayBase<A, Ix2>,
+    w: &Array2<F>, // This parameter includes `W` and `b`
+) -> (Array2<F>, Array2<F>) {
+    let n_features = x.shape()[1];
+    let (params, intercept) = convert_params(n_features, w);
+    // Compute H
+    let h = x.dot(&params) + intercept;
+    // This computes `H - log(sum(exp(H)))`, which is equal to
+    // `log(softmax(H)) = log(exp(H) / sum(exp(H)))`
+    let log_prob = &h - log_sum_exp(&h, Axis(1)).into_shape((h.nrows(), 1)).unwrap();
+    (log_prob, params)
+}
+
+/// Computes loss function of `-sum(Y * log(softmax(H))) + alpha/2 * norm(W)`
+fn multi_logistic_loss<F: Float, A: Data<Elem = F>>(
+    x: &ArrayBase<A, Ix2>,
+    y: &Array2<F>,
+    alpha: F,
+    w: &Array2<F>,
+) -> F {
+    let (log_prob, params) = multi_logistic_prob_params(x, w);
+    // Calculate loss
+    // XXX Should we divide cost by n_samples???
+    -elem_dot(&log_prob, y) + F::cast(0.5) * alpha * elem_dot(&params, &params)
```

Comment:
because this is an estimate of the chained likelihood (and not MSE), I would not divide by the sample size

Reply:
on the other hand this is only true if the chain has always the same length otherwise the `alpha` would vary with the sample size